spring-content-rest/src/main/java/internal/org/springframework/content/rest/utils/ContentStoreUtils.java

```diff
@@ -1,83 +1,146 @@
-		
+
+		// content name
+		Field contentName = BeanUtils.getFieldWithAnnotationField(entity,
+				org.springframework.content.commons.annotations.ContentName.class);
+		if (contentName != null) {
+			Annotation cntNameAn = contentName
+					.getAnnotation(org.springframework.content.commons.annotations.ContentName.class);
+			cName = BeanUtils.getFieldValue(entity, contentName).toString();
+			if (cntNameAn != null) {
+				httpNameHeader = AnnotationUtils.getValue(cntNameAn, "httpHeader").toString();
+			}
+		}
+
 		MediaType targetMimeType = MediaType.valueOf(entityMimeType.toString());
-		
+
 		MediaType.sortBySpecificityAndQuality(mimeTypes);
-		
+
 		MediaType[] arrMimeTypes = mimeTypes.toArray(new MediaType[] {});
-		
-		//Modified to show download
+
+		// Modified to show download
 		Object originalFileName = BeanUtils.getFieldWithAnnotation(entity, OriginalFileName.class);
 		if (originalFileName != null) {
-			headers.setContentDispositionFormData("attachment", (String)originalFileName);
+			headers.setContentDispositionFormData("attachment", (String) originalFileName);
 		}
-		
-		for (int i=0; i < arrMimeTypes.length && content == null; i++) {
+
+		for (int i = 0; i < arrMimeTypes.length && content == null; i++) {
 			MediaType mimeType = arrMimeTypes[i];
-			if (mimeType.includes(targetMimeType)) {
+			if (!(mimeType.isConcrete() && targetMimeType.isConcrete()
+					&& mimeType.getType().equals(targetMimeType.getType())
+					&& mimeType.getSubtype().equals(targetMimeType.getSubtype()) && !mimeType.equals(targetMimeType)) && // If
+																															// the
+																															// same
+																															// but
+																															// with
+																															// parameters,
+																															// use
+																															// renderer...
+					mimeType.includes(targetMimeType)) { // ... or load original in compatibility mode
 				headers.setContentType(targetMimeType);
-				
-				long contentLength = 0L;
+
+				// content name header
+				headers.set(httpNameHeader, cName);
+
+				// long contentLength = 0L;
 				Object len = BeanUtils.getFieldWithAnnotation(entity, ContentLength.class);
 				if (len != null)
 					headers.setContentLength(Long.parseLong(len.toString()));
-				
+
 				content = store.getContent(entity);
 				break;
 			} else if (store instanceof Renderable) {
-				content = ((Renderable<Object>)store).getRendition(entity, mimeType.toString());
+				Resource ret = ((Renderable<Object>) store).getRendition(entity, mimeType.toString());
+				if (ret != null) {
+					// we have to use mime arrived from rendition due to it can be one of possible
+					// requested
+					String mt = (ret instanceof MedializedResource) ? ((MedializedResource) ret).getMime()
+							: "application/octet-stream";
+					headers.setContentType(MediaType.valueOf(mt));
+
+					// determine file extension TODO: make it better!!!
+					String mtExt = ".new";
```

Comment:
I am guessing you want to add an extension so that the client see it as "different" from the original content?

Reply:
After translation I need new file name extension, In my code I have used apache tika, so in this code I put just TODO

Reply:
Sure.  Makes sense.  My concern with just ".new" is what happens if your client calls for different renditions of that content?  Would a non-unique file and extension cause the client to think it has a rendition cached?  Anyways, it's TODO so let's just figure it out as we move forwards.  I understand why it is there.

Reply:
In my project code this part is solved with sort of static map mime->extension. initially we can use the same, should we?